### PR TITLE
Add docker-exec command

### PR DESCRIPTION
**Pull request by wkolean**
Added docker-exec [options] container -- [command] which invokes `docker exec` to run a command in a running container #4904 #4179.

If the container is not running, docker-exec will start the container and leave the container running after the command has finished. This makes it easy to start a container and go directly into an interactive shell using one command `vagrant docker-exec -t app -- bash`

Also added a command for opening a shell, `vagrant docker-shell app` #4179 which is just a shortcut for `vagrant docker-exec -t app -- bash`

I basically hacked docker-run and docker-logs to get this to work, so if any of the code doesn't make sense then it's probably because I didn't fully understand what I was copying.

This has been tested on a Mac running VirtualBox and using vagrant_vagrantfile to define a phusion/ubuntu-14.04-amd64 box. It was also tested against containers that had both images and Dockerfiles. Command doesn't work using the internal boot2docker because the installed docker version is too old. There already is a pull request to update boot2docker, and this PR probably shouldn't merge until boot2docker is updated first.

Note that docker-run doesn't mount any shared folders from config.vm.synced_folder so docker-exec doesn't create synced folders either (except for /vagrant which still works). I think this is a bug, but I haven't found a solution. If I find one, I'll create a separate PR.

### CODE REVIEW

File: plugins/providers/docker/driver.rb [hunk 1]
```diff
@@ -151,6 +151,17 @@ def docker_bridge_ip
         end
       end
 
+      def exec(params, **opts, &block)
+        cmd      = Array(params.fetch(:cmd))
+        exec_cmd = %W(docker exec)
```
Comment (sethvargo) [inline plugins/providers/docker/driver.rb]: We don't need interpolation, so lowercase `w` should be fine here

File: plugins/providers/docker/driver.rb [hunk 2]
```diff
@@ -151,6 +151,17 @@ def docker_bridge_ip
         end
       end
 
+      def exec(params, **opts, &block)
```
Comment (sethvargo) [inline plugins/providers/docker/driver.rb]: We are moving away from keyword args and using `opts = {}` instead. The keyword args require a specific version of Ruby and only work for symbols.

File: plugins/providers/docker/command/exec.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        end
+
+        def execute
+          options = {}
+          options[:detach] = false
+          options[:pty] = false
+          options[:rm] = false
+
+          opts = OptionParser.new do |o|
+            o.banner = "Usage: vagrant docker-exec [command...]"
```
Comment (sethvargo) [inline plugins/providers/docker/command/exec.rb]: I think this is actually

```
Usage: vagrant docker-exec [options] -- command
```

right?

File: plugins/providers/docker/command/exec.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+            machine.action(
+              :exec_command,
+              exec_command: command,
+              exec_detach: options[:detach],
+              exec_pty: options[:pty],
+              exec_rm: options[:rm]
+            )
+          end
+
+          0
```
Comment (sethvargo) [inline plugins/providers/docker/command/exec.rb]: An explicit `return` here would be nice :smile:

**Comment by sethvargo** [comment]
Hi @wkolean 

Thank you for the Pull Request, and I apologize for the delay in reviewing. This looks _really_ good. I made a few style comments and added some questions. Please let me know if anything is unclear.

Finally, it would be great to have some test coverage around this. I know most Vagrant plugins are not well tested, but if take a look at the Chef provisioner or pushes, there is some decent test coverage/skeleton you can copy :smile:.

**Comment by wkolean** [comment]
Thanks for the great feedback! I'll make the changes, and there is a stand-alone plugin I created with some updates that I'll port over. Adding tests is a good idea, thank you for the pointers.

File: plugins/providers/docker/driver.rb (hunk 2 above)
Comment (wkolean) [inline plugins/providers/docker/driver.rb]: The docker provider makes heavy use of **opts, I'm not sure how to modify just my changes to not use **opts without refactoring the the rest of the plugin

**Comment by thomaszbz** [comment]
@sethvargo @wkolean This feature is so cool. Are you still working on this? Would like to see it released soon.

**Pull request closed without merging**